lib/pusher/native_notification/client.rb

```diff
@@ -0,0 +1,123 @@
+module Pusher
+  module NativeNotification
+    class Client
+      attr_reader :app_id, :host
+
+      API_PREFIX = "customer_api"
+      API_VERSION = "v1"
+      GCM_TTL = 241920
+      RESTRICTED_GCM_PAYLOAD_KEYS = [:to, :registration_ids]
+      WEBHOOK_LEVELS = ["DEBUG", "INFO"]
+
+      def initialize(app_id, host, scheme, pusher_client)
+        @app_id = app_id
+        @host = host
+        @scheme = scheme
+        @pusher_client = pusher_client
+      end
+
+      # Send a notification via the native notifications API
+      def notify(interests, data = {})
+        Request.new(
+          @pusher_client,
+          :post,
+          url("/notifications"),
+          {},
+          payload(interests, data)
+        ).send_sync
+      end
+
+      private
+
+      # TODO: Actual links
+      #
+      # {
+      #   interests: [Array of interests],
+      #   apns: {
+      #     See https://pusher.com/docs/native_notifications/payloads#apns
+      #   },
+      #   gcm: {
+      #     See https://pusher.com/docs/native_notifications/payloads#gcm
+      #   }
+      # }
+      #
+      # @raise [Pusher::Error] if the `apns` or `gcm` key does not exist
+      # @return [String]
+      def payload(interests, data)
+        interests = Array(interests).map(&:to_s)
+
+        raise Pusher::Error, "Too many interests provided" if interests.length > 1
+
+        data = deep_symbolize_keys!(data)
+        validate_payload(data)
+
+        data.merge!(interests: interests)
+
+        MultiJson.encode(data)
+      end
+
+      def url(path = nil)
+        URI.parse("#{@scheme}://#{@host}/#{API_PREFIX}/#{API_VERSION}/apps/#{@app_id}#{path}")
+      end
+
+      # Validate payload
+      # `time_to_live` -> value b/w 0 and 241920
+      # If the `notification` key is provided, ensure
+      # that there is an accompanying `title` and `icon`
+      # field
+      def validate_payload(payload)
+        unless (payload.has_key?(:apns) || payload.has_key?(:gcm))
+          raise Pusher::Error, "GCM or APNS data must be provided"
+        end
+
+        if (gcm_payload = payload[:gcm])
+          # Restricted keys
+          RESTRICTED_GCM_PAYLOAD_KEYS.each { |k| gcm_payload.delete(k) }
+          if (ttl = gcm_payload[:time_to_live])
+
+            if ttl.to_i < 0 || ttl.to_i > GCM_TTL
+              raise Pusher::Error, "Time to live must be between 0 and 241920 (4 weeks)"
```

Comment:
There is validation on the server for all cases. Probably better to validate both places?